File: shopcaovanson/services/auth-service/internal/domain/roles.go
package domain

const (
	RoleSuperAdmin = "super_admin"
	RoleAdmin      = "admin"
	RoleManager    = "manager"
	RoleStaff      = "staff"
	RoleSupport    = "support"
	RoleCustomer   = "customer"
)

var StaffRoles = []string{
	RoleSuperAdmin,
	RoleAdmin,
	RoleManager,
	RoleStaff,
	RoleSupport,
}

var roleLevel = map[string]int{
	RoleSuperAdmin: 100,
	RoleAdmin:      80,
	RoleManager:    60,
	RoleStaff:      40,
	RoleSupport:    20,
	RoleCustomer:   0,
}

func RoleLevel(role string) int {
	if v, ok := roleLevel[role]; ok {
		return v
	}
	return 0
}

func HasMinRole(role, minRole string) bool {
	return RoleLevel(role) >= RoleLevel(minRole)
}

func IsStaffRole(role string) bool {
	return RoleLevel(role) >= RoleLevel(RoleSupport)
}

func CanManageUsers(role string) bool {
	return RoleLevel(role) >= RoleLevel(RoleAdmin)
}

func CanAssignRole(actorRole, targetRole string) bool {
	return RoleLevel(actorRole) > RoleLevel(targetRole)
}

func ValidStaffRoles() []string {
	return StaffRoles
}

File: shopcaovanson/frontend/web/utils/roles.ts
export type UserRole =
  | 'super_admin'
  | 'admin'
  | 'manager'
  | 'staff'
  | 'support'
  | 'customer'

const roleLevel: Record<UserRole, number> = {
  super_admin: 100,
  admin: 80,
  manager: 60,
  staff: 40,
  support: 20,
  customer: 0,
}

export function roleLevelOf(role: string): number {
  return roleLevel[role as UserRole] ?? 0
}

export function hasMinRole(role: string, min: UserRole): boolean {
  return roleLevelOf(role) >= roleLevel[min]
}

export const STAFF_ROLES: UserRole[] = ['super_admin', 'admin', 'manager', 'staff', 'support']
